Added xSPSearchCrawlRule resource

ykuijs opened the pull request:
Added a new resource xSPSearchCrawlRule to manage Enterprise Search crawl rules as mentioned in issue #29 

Thanks for submitting a Pull Request to this project.

Please make sure you have read the [Contribution Guidelines](https://github.com/powershell/xSharePoint/wiki/Contributing%20to%20xSharePoint).

To aid community reviewers in reviewing and merging your PR, please take the time to run through this checklist:
- [X] Change details added to Unreleased section of readme.md?
- [X] Added/updated documentation and descriptions in .schema.mof files where appropriate?
- [X] Examples updated for both the single server and small farm templates in the examples folder?
- [X] New/changed code adheres to [Style Guidelines]?(https://github.com/PowerShell/DscResources/blob/master/StyleGuidelines.md)?
- [X] [Unit and Integration tests](https://github.com/PowerShell/DscResources/blob/master/TestsGuidelines.md) created/updated where possible?

Your contribution to this project is greatly appreciated.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/xsharepoint/290)
<!-- Reviewable:end -->

BrianFarnhill commented:
Looks like it's failing at AppVeyor on the PSScriptAnalyser and some tests on the schema. I'll pick some of that up in the review below but definitely needs some attention. I didn't get to the review the tests yet, but I submitted the stuff I have so far so you could begin on that and I'll review the tests at the next revision. Thanks bud.

---

Reviewed 5 of 6 files at r1.
Review status: 5 of 6 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[README.md, line 126 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGilaWf-u3zPFKVbgAb:-KGilaWgkWVmSDQGUWj0:-1709040928) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/README.md#L126)):*
There is a line above that introduces the new resources for this release, can you add this to the existing line?

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 22 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGilvqCIzub_8FE6MJE:-KGilvqCIzub_8FE6MJF:-1090796165) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L22)):*
Can you elaborate for me on why we can't remove rules of those types?

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 23 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGioVGWHz7cODwet5ht:-KGioVGWHz7cODwet5hu:1731759062) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L23)):*
Why are we returning null here? If there is an issue with the parameters that stop us from returning the current state we should be throwing exceptions here to indicate something stopped us from retrieving the current state, but if the invalid presentation of parameters won't stop the get call we should be letting this go through and throw the exception when we try to test or set the resource. The same applies for all the return null lines below this in the get method.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 28 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGimKDSqQMUwMYkqnNW:-KGimKDTATlPiSkkwudS:-1063692439) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L28)):*
Can we change this message to be a bit clearer - something like "If AuthenticationType is CertificateRuleAccess you must specify the CertificateName value"

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 34 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGimXCXKoRCJVfNk-qw:-KGimXCXKoRCJVfNk-qx:1188937699) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L34)):*
Same as above for more specific details in the error - "when X=Y then you must provide value Z" kind of thing.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 40 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGimckBm2k7RjWMLwCs:-KGimckBm2k7RjWMLwCt:-1619434713) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L40)):*
And again on error message formatting here

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 46 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGimfod-7fInfeKq0La:-KGimfoeuHb822JtK1Ws:923002971) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L46)):*
Error message formatting

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 52 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGimidY76YVmzDp8bfw:-KGimidY76YVmzDp8bfx:923002971) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L52)):*
Error message formatting

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 96 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGinMcbn-w4ULusSjJY:-KGinMccHhVMmdrX0GKN:-523356482) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L96)):*
This looks like it will return the [System.Type] object back here, where as I would assume we want a string here right? Should this then be $crawlRule.Type.Name

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 108 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGinlF7aUtjtg-uGmCe:-KGinlF7aUtjtg-uGmCf:813782705) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L108)):*
As above with the Type object and its name property

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 116 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGin95bbrFFO8YvulIF:-KGin95bbrFFO8YvulIG:-319412009) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L116)):*
Rather than use default here can we specify the types you expect to catch here. This way if the product is ever changed our code won't inherit a new type that it isn't expecting

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 121 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGinoESamxxBFJbOqnY:-KGinoESamxxBFJbOqnZ:1849682007) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L121)):*
The ToString() here will return the full name I think rather than just the short name which is what I think you are expecting. Either way if you could make this output the same as the two above it for handling the Type value that would be better

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 153 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGiosw5qcjlGr8nk6t4:-KGiosw6q5deiGVSVRkv:-640078318) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L153)):*
We shouldn't be returning $null from the get method, so we might need to refactor how you do your check for parameters being valid. 

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 156 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGiptRWLtjMcNHFvojl:-KGiptRWLtjMcNHFvojm:-126263290) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L156)):*
Just from the point of view of cleaning up the code, we can probably ditch the comment here seeing as we have a write-verbose line right below it. Same for the other two below this in the set method

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 175 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGipJozsn21OzBkwfJU:-KGipJozsn21OzBkwfJV:1948573633) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L175)):*
You can just use $params.AuthenticationCredentials.Password to get the secure string here rather than taking it out as plain text and then back to a secure string. The Script Analyser will complain about this as well because it will put the password in to memory as plain text temporarily so we need to avoid this (my PR to move away from experimental fixes this in a bunch of places as well so the other resources we do this for won't be that way for long)

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 205 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGipjb9YiKQis3QzhQs:-KGipjbA1nqBws45Dv--:-847350284) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L205)):*
Same note here about the secure string of the password

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.schema.mof, line 4 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGiqBucjBu7NQ4wK8wl:-KGiqBudD83z-gZnxlVB:-205775172) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.schema.mof#L4)):*
Looks like the description needs to be updated from what you copied from here

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.schema.mof, line 39 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGiqQB2a4JS0lwZXt-Q:-KGiqQB2a4JS0lwZXt-R:550740469) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.schema.mof#L39)):*
Can we change this to "RuleType" - the develoepr in me knows that Type is a property on way too many object types in .NET and I want to make sure we avoid using reserved names for things. I don't think this will cause a specific issue right now but might head off something in future. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-:-KGil8pWS6vACrCAqgxr:-937418773)*
<!-- Sent from Reviewable.io -->

ykuijs commented:
Review status: 3 of 6 files reviewed at latest revision, 18 unresolved discussions.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 22 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGilvqCIzub_8FE6MJE:-KGlv1mWyQZd_eQhBwxN:-729125597) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L22)):*
The Ensure=Absent matches on the rule name and ServiceApp name. Since it does not make sense to specify additional parameters if you want to delete the rule, I build in this check.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 23 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGioVGWHz7cODwet5ht:-KGlvs1O2KqAi9-TIZVw:-1282906015) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L23)):*
Ok, will change that. I saw the same behavior in other resources so I copied that.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 28 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGimKDSqQMUwMYkqnNW:-KGlw-aE5diPyi4on1om:674544752) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L28)):*
Will try to improve the message. This check checks if either one of the parameters is present without the other being present.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 34 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGimXCXKoRCJVfNk-qw:-KGlwfuYzYrCIdTpgun7:674544752) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L34)):*
Will try to improve the message. This check checks if either one of the parameters is present without the other being present.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 40 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGimckBm2k7RjWMLwCs:-KGlwii29Cyyr4kT9iV3:674544752) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L40)):*
Will try to improve the message. This check checks if either one of the parameters is present without the other being present.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 46 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGimfod-7fInfeKq0La:-KGlwjt6W87WyiEVj8oM:674544752) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L46)):*
Will try to improve the message. This check checks if either one of the parameters is present without the other being present.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 52 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGimidY76YVmzDp8bfw:-KGlwlG1hoEFW7CarOuM:674544752) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L52)):*
Will try to improve the message. This check checks if either one of the parameters is present without the other being present.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 96 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGinMcbn-w4ULusSjJY:-KGlwooywOdB9Ro5y6X3:-1251014914) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L96)):*
Type is a property of the object. I think you are confused with the GetType() method

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 108 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGinlF7aUtjtg-uGmCe:-KGlx34RBLi_dZGBQ0qZ:-1251014914) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L108)):*
Type is a property of the object. I think you are confused with the GetType() method

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 116 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGin95bbrFFO8YvulIF:-KGlx9L09l6-g33O2d0x:2133327276) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L116)):*
Ok, will update.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 121 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGinoESamxxBFJbOqnY:-KGlxEQou1FwBG6vu4ig:1286356735) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L121)):*
Type is a property of the object. I think you are confused with the GetType() method. Will change the ToString() to match all three instances.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 153 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGiosw5qcjlGr8nk6t4:-KGlxPKk6NcqNRwDXGuD:866299604) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L153)):*
Will do

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 156 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGiptRWLtjMcNHFvojl:-KGlxWBCxCPD520-X5es:866299604) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L156)):*
Will do

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.schema.mof, line 4 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGiqBucjBu7NQ4wK8wl:-KGlxl3xdVlI84H931Wn:1895572966) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.schema.mof#L4)):*
Thought I changed this one, will update :-)

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.schema.mof, line 39 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGiqQB2a4JS0lwZXt-Q:-KGlxxUSOdQOLQCRBuHe:1085483980) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.schema.mof#L39)):*
Will do 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsharepoint/290)*
<!-- Sent from Reviewable.io -->

ykuijs commented:
Review status: 3 of 6 files reviewed at latest revision, 16 unresolved discussions.

---

*[README.md, line 126 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGilaWf-u3zPFKVbgAb:-KGmTNRYGxBoFlxHpFzL:-499146009) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/README.md#L126)):*
Done.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 22 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGilvqCIzub_8FE6MJE:-KGmas9pmrQL-ylMrZri:-499146009) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L22)):*
Done.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 23 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGioVGWHz7cODwet5ht:-KGmTTGFg33XdpYjhIFm:-499146009) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L23)):*
Done.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 28 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGimKDSqQMUwMYkqnNW:-KGmTXHbOodILpVVaZIl:-499146009) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L28)):*
Done.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 34 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGimXCXKoRCJVfNk-qw:-KGmT_IX8BtoTFQIXREc:-499146009) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L34)):*
Done.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 40 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGimckBm2k7RjWMLwCs:-KGmTc-FxPbq3SV7neTv:-499146009) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L40)):*
Done.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 46 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGimfod-7fInfeKq0La:-KGmTeFQdcJXN1UhQci2:-499146009) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L46)):*
Done.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 52 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGimidY76YVmzDp8bfw:-KGmThXAX-g5ujtcKrtU:-499146009) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L52)):*
Done.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 116 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGin95bbrFFO8YvulIF:-KGmVaF2gANFHZOtI6dQ:-499146009) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L116)):*
Done.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 121 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGinoESamxxBFJbOqnY:-KGmVdjQs-b8gl4KPVRQ:-499146009) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L121)):*
Done.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 153 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGiosw5qcjlGr8nk6t4:-KGmVl6iX2Ivun1Lo54k:-499146009) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L153)):*
Done.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 156 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGiptRWLtjMcNHFvojl:-KGmVui2tOg7Uvf94ip8:-499146009) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L156)):*
Done.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 175 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGipJozsn21OzBkwfJU:-KGmW-DGPIS-WFRV_MqE:1518467758) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L175)):*
Removed the ConvertTo-SecureString, since the cmdlet requires a plaint text password

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 205 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGipjb9YiKQis3QzhQs:-KGmWNnXs6MdUirjEwnw:1518467758) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L205)):*
Removed the ConvertTo-SecureString, since the cmdlet requires a plaint text password

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.schema.mof, line 4 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGiqBucjBu7NQ4wK8wl:-KGmWr9NQqsy50es8FLr:-499146009) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.schema.mof#L4)):*
Done.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.schema.mof, line 39 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGiqQB2a4JS0lwZXt-Q:-KGmadZofWiuSj38rHeT:-499146009) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.schema.mof#L39)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsharepoint/290)*
<!-- Sent from Reviewable.io -->

ykuijs commented:
Testing all changes before pushing the commit

ykuijs commented:
Pester tests now succeed, needed to sync commits to be able to test on my test environment. All review comments have been incorporated.

BrianFarnhill commented:
Good review mate, few more little things to touch up to get this over the line.

---

Reviewed 3 of 4 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 23 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGioVGWHz7cODwet5ht:-KGnOybBwXODucOGcqi1:641780880) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L23)):*
Yea I'm on a cleanup mission for stuff like this at the moment, I know I've got most of them sorted but not all yet. Thanks :)

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 156 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGiptRWLtjMcNHFvojl:-KGnROhSinKzSzzvBQ4o:-1981626442) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L156)):*
You tagged this as done but I still see both the comment and the verbose output here?

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 175 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGipJozsn21OzBkwfJU:-KGnRZR-8UrI7uNXOnwr:-914253272) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L175)):*
If the cmdlet needs plain text you'll actually need to make this $params.AuthenticationCredentials.GetNetworkCredential().Password - the straight password property actually sends a secure string. 

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 205 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGipjb9YiKQis3QzhQs:-KGnRjIpYG6LIZIGEHO_:257153793) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L205)):*
Same note here about the secure string

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-:-KGnS0rqgVzmvwaJA80b:-730485566)*
<!-- Sent from Reviewable.io -->

BrianFarnhill commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 175 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGipJozsn21OzBkwfJU:-KGnSQOpztzKhTe2nTNf:-497824149) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L175)):*
Just updating quickly because I tagged this as resolved when it needs a change

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 205 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGipjb9YiKQis3QzhQs:-KGnSWDwaIdsTe7Apuop:1854746597) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L205)):*
Same here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsharepoint/290)*
<!-- Sent from Reviewable.io -->

ykuijs commented:
Changed the comments. Doing a new commit now

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 156 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGiptRWLtjMcNHFvojl:-KH1HK5NrFJEzdk1hIa5:-90048427) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L156)):*
Done, changed this on a different line. Now did in in the entire file :-)

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 175 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGipJozsn21OzBkwfJU:-KH1HUCuS2avqbYqgOfO:670530871) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L175)):*
The cmdlets indeed needs a plain text password. However during testing this resulted in an error indicating that the cmdlet required a secure string.  That is why I now changed it back to the secure string, but then the option you suggested earlier.

---

*[Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1, line 205 \[r1\]](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-KGipjb9YiKQis3QzhQs:-KH1HwyAHmWrVpsa_MX7:670530871) ([raw file](https://github.com/powershell/xsharepoint/blob/ce0107e1b366e2cde197ece876c220aaf42cf1a3/Modules/xSharePoint/DSCResources/MSFT_xSPSearchCrawlRule/MSFT_xSPSearchCrawlRule.psm1#L205)):*
The cmdlets indeed needs a plain text password. However during testing this resulted in an error indicating that the cmdlet required a secure string.  That is why I now changed it back to the secure string, but then the option you suggested earlier.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-:-KH1I22jFazEM-qE2E3y:-1031837253)*
<!-- Sent from Reviewable.io -->

BrianFarnhill commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 5 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsharepoint/290#-:-KH1t0nHEV3sOEDrqVfg:1417088721)*
<!-- Sent from Reviewable.io -->

BrianFarnhill commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xsharepoint/290)*
<!-- Sent from Reviewable.io -->
